src/test/scala/edu/ie3/simona/model/participant/StorageModelSpec.scala

```diff
@@ -0,0 +1,418 @@
+          "expActiveNext",
+          "expScheduled",
+          "expDelta",
+        ),
+        // no power
+        (0.0, 0.0, 0.0, false, false, 0),
+        (50.0, 0.0, 0.0, false, false, 0),
+        (100.0, 0.0, 0.0, false, false, 0),
+        // charging on empty
+        (0.0, 1.0, 1.0, true, true, (50 * 3600 / 0.9).toInt),
+        (0.0, 2.5, 2.5, true, true, (20 * 3600 / 0.9).toInt),
+        (0.0, 5.0, 5.0, true, true, (10 * 3600 / 0.9).toInt),
+        (0.0, 10.0, 10.0, true, true, (5 * 3600 / 0.9).toInt),
+        // charging on target ref
+        (50.0, 5.0, 5.0, true, true, (10 * 3600 / 0.9).toInt),
+        (50.0, 10.0, 10.0, true, true, (5 * 3600 / 0.9).toInt),
+        // discharging on target ref
+        (50.0, -4.5, -4.5, true, true, 10 * 3600),
+        (50.0, -9.0, -9.0, true, true, 5 * 3600),
+        // discharging on full
+        (100.0, -4.5, -4.5, true, true, 10 * 3600),
+        (100.0, -9.0, -9.0, true, true, 5 * 3600),
+      )
+
+      forAll(testCases) {
+        (
+            lastStored: Double,
+            setPower: Double,
+            expPower: Double,
+            expActiveNext: Boolean,
+            expScheduled: Boolean,
+            expDelta: Int,
+        ) =>
+          val oldState = StorageModel.StorageState(
+            KilowattHours(lastStored),
+            Kilowatts(0d),
+            startTick,
+          )
+
+          val result = storageModel.handleControlledPowerChange(
+            data,
+            oldState,
+            Kilowatts(setPower),
+          )
+
+          val (newState, flexChangeIndication) = result
+
+          newState.chargingPower should approximate(Kilowatts(expPower))
+          newState.tick shouldBe (startTick + 1)
+          newState.storedEnergy should approximate(KilowattHours(lastStored))
+
+          flexChangeIndication.changesAtTick.isDefined shouldBe expScheduled
+          flexChangeIndication.changesAtTick.forall(
+            _ == (startTick + 1 + expDelta)
+          ) shouldBe true
+          flexChangeIndication.changesAtNextActivation shouldBe expActiveNext
+      }
+    }
+
+    "Handle the edge case of discharging in tolerance margins" in {
+      val storageModel = buildStorageModel()
+      val startTick = 1800L
+      val data = StorageModel.StorageRelevantData(startTick + 1)
+      // margin is at ~ 0.0030864 kWh
+      val oldState = StorageModel.StorageState(
+        KilowattHours(0.002d),
+        Kilowatts(0d),
+        startTick,
+      )
+
+      val result = storageModel.handleControlledPowerChange(
+        data,
+        oldState,
+        Kilowatts(-5d),
+      )
+
+      result._1.chargingPower should approximate(Kilowatts(0))
+      result._1.tick shouldBe (startTick + 1)
+      result._1.storedEnergy should approximate(oldState.storedEnergy)
+
+      val flexChangeIndication = result._2
+      flexChangeIndication.changesAtTick.isDefined shouldBe false
+      flexChangeIndication.changesAtNextActivation shouldBe true
```

Comment:
If you change 'val result = .... '  into 'val (newState, flexChangeIndication) = storageModel.handleControlledPowerChange(...) '

this can improved in terms of readabliity